# Lack of Failsafe Mechanism in the withdraw() function

## Finding description and impact

The withdraw() function does not have a check to prevent withdrawing titn tokens meant for users.

**Why Does the Function Lack a Failsafe Mechanism?**

The `withdraw` function allows the owner to withdraw any amount of any ERC20 token from the contract without any checks or safeguards. Specifically:
1. It does not validate the `amount` parameter to ensure it is within reasonable bounds.
2. It does not check whether the contract has sufficient balance of the token being withdrawn.
3. It does not restrict the tokens that can be withdrawn (e.g., only allowing `TGT` or `TITN`).

**Loss of Funds**:
   - If the owner accidentally withdraws too many tokens or the wrong token, the contract could lose funds that are critical to its operation.


## Proof of Concept

https://github.com/code-423n4/2025-02-thorwallet/blob/98d7e936518ebd80e2029d782ffe763a3732a792/contracts/MergeTgt.sol#L59-L62

```solidity
    function withdraw(IERC20 token, uint256 amount) external onlyOwner {
        token.safeTransfer(owner(), amount);
        emit Withdraw(address(token), amount, owner());
    }
```


An attacker or an external user could call the function with a token contract that returns false upon transfer, leading to a failed transaction without reverting. In such cases, no Withdraw event would be emitted, but the state of the contract would remain unaltered.

If an incorrect amount is passed (e.g., more than the available balance), the transfer will fail, and the transaction will revert without any specific error handling.


## Recommended mitigation steps

Implement checks to ensure that the transfer will be successful. This can include checking the balance of the contract, the allowance, and verifying the success of the transfer.

Use a require statement to ensure that the token transfer was successful.

